xwiki-commons-core/xwiki-commons-job/src/main/java/org/xwiki/job/internal/JobGroupPathLockTree.java

```diff
@@ -32,31 +38,104 @@
  * @version $Id$
  * @since 6.1M2
  */
+@Component(roles = JobGroupPathLockTree.class)
+@Singleton
 public class JobGroupPathLockTree
 {
-    private final Map<JobGroupPath, ReadWriteLock> tree = new ConcurrentHashMap<>();
+    @Inject
+    private GroupedJobInitializerManager groupedJobInitializerManager;
+
+    /**
+     * A specific concurrency implementation for managing Semaphore with Read/Write lock capabilities.
+     */
+    private class ReadWriteSemaphore
+    {
+        private volatile int readCounter;
+        private volatile int writeCounter;
+        private Semaphore semaphore;
+        private Lock writeCounterLock;
+
+        ReadWriteSemaphore(int poolSize)
+        {
+            this.semaphore = new Semaphore(poolSize, true);
+            this.writeCounterLock = new ReentrantLock(true);
+            this.readCounter = 0;
+            this.writeCounter = 0;
+        }
+
+        void lockWrite()
+        {
+            this.writeCounterLock.lock();
+            this.writeCounter++;
+            this.writeCounterLock.unlock();
+            this.semaphore.acquireUninterruptibly(this.readCounter + 1);
+        }
+
+        void unlockWrite()
+        {
+            this.writeCounterLock.lock();
+            this.writeCounter--;
+
+            if (this.writeCounter == 0) {
+                this.writeCounterLock.unlock();
```

Comment:
You have the exact same line in both branches of the if so I guess it could be before it.